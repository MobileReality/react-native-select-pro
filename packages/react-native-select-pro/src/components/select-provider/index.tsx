import React, { ReactNode } from 'react';
import { PortalHost, PortalProvider } from '@gorhom/portal';

import { Portals } from '../../constants/portals';

type Props = {
    children: ReactNode;
};

const PortalHosts = () => {
    return (
        <>
            <PortalHost name={Portals.SelectOutsideWrapper} />
            <PortalHost name={Portals.Select} />
        </>
    );
};

export const SelectProvider = ({ children }: Props) => {
    return (
        <PortalProvider>
            {children}
            <PortalHosts />
        </PortalProvider>
    );
};

export const SelectModalProvider = ({ children }: Props) => {
    return (
        <>
            <PortalHosts />
            {children}
        </>
    );
};
