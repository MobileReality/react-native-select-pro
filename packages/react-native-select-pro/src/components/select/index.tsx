import React, {
    ForwardedRef,
    forwardRef,
    RefObject,
    useEffect,
    useImperativeHandle,
    useReducer,
    useRef,
} from 'react';
import {
    I18nManager,
    StyleSheet,
    TextInput,
    useWindowDimensions,
    View,
    ViewStyle,
} from 'react-native';

import {
    ANIMATION_DURATION,
    COLORS,
    ITEM_HEIGHT,
    MAX_HEIGHT_LIST,
} from '../../constants/styles';
import { getSize } from '../../helpers';
import { initialData, reducer } from '../../state/reducer';
import { Action } from '../../state/types';
import type {
    OnOutsidePress,
    OnPressOptionType,
    OnPressSelectControlType,
    OptionType,
    SelectProps,
    SelectRef,
} from '../../types';
import { OptionsList } from '../options-list';
import { SelectControl } from '../select-control';

export const Select = forwardRef(
    (props: SelectProps, ref: ForwardedRef<SelectRef>) => {
        const {
            // Required
            options,
            // Callbacks
            onSelect,
            onDropdownOpened,
            onDropdownClosed,
            // Texts
            noOptionsText = 'No options',
            placeholderText = 'Select...',
            // Animations
            animated = false,
            animationDuration = ANIMATION_DURATION,
            // Behaviour
            clearable = true,
            closeDropdownOnSelect = true,
            disabled = false,
            scrollToSelectedOption = true,
            hideSelectControlArrow = false,
            // Additional features
            defaultOption,
            flatListProps,
            // Search
            searchable = false,
            searchPattern = (payload: string) => `(${payload})`,
            // Multiselect
            multiSelection = false,
            // Custom components
            NoOptionsComponent,
            OptionComponent,
            // Custom sources
            customLeftIconSource,
            // Colors
            placeholderTextColor = COLORS.GRAY,
            // Accessibility
            selectControlClearOptionA11yLabel,
            selectControlOpenDropdownA11yLabel,
            // Styles
            optionStyle,
            optionsListStyle,
            optionTextStyle,
            optionSelectedStyle,
            selectContainerStyle,
            selectControlArrowImageStyle,
            selectControlButtonsContainerStyle,
            selectControlClearOptionButtonStyle,
            selectControlClearOptionImageStyle,
            selectControlClearOptionButtonHitSlop,
            selectControlDisabledStyle,
            selectControlStyle,
            selectControlTextStyle,
            customLeftIconStyles,
            multiSelectionOptionStyle,
            inputProps,
        } = props;
        const [state, dispatch] = useReducer(reducer, initialData);
        const {
            isOpened,
            selectedOption,
            optionsData,
            openedPosition,
            searchValue,
            searchedOptions,
            searchInputRef,
            selectedOptionIndex,
        } = state;
        const { aboveSelectControl } = openedPosition;
        const selectedOptionTyped = selectedOption as OptionType;

        const containerRef = useRef<View>(null);

        useEffect(() => {
            if (!Array.isArray(options)) {
                // eslint-disable-next-line no-console
                console.error('You must pass array in the options prop');
                return;
            }

            if (options.length > 0) {
                dispatch({ type: Action.SetOptionsData, payload: options });

                const isValidPassDefaultOption =
                    defaultOption &&
                    // eslint-disable-next-line no-prototype-builtins
                    defaultOption.hasOwnProperty('value') &&
                    // eslint-disable-next-line no-prototype-builtins
                    defaultOption.hasOwnProperty('label');

                if (isValidPassDefaultOption) {
                    const foundIndex = options.findIndex(
                        ({ value }) => value === defaultOption.value,
                    );
                    dispatch({
                        type: Action.SelectOption,
                        payload: {
                            selectedOption: defaultOption,
                            selectedOptionIndex: foundIndex,
                        },
                    });
                }
            }
        }, [options]);

        useImperativeHandle(ref, () => ({
            clear: () => {
                dispatch({
                    type: Action.SelectOption,
                    payload: { selectedOption: null, selectedOptionIndex: -1 },
                });
                dispatch({ type: Action.SetOptionsData, payload: options });
            },
            open: () => {
                if (containerRef.current && !disabled) {
                    dispatch({
                        type: Action.Open,
                    });
                    setPosition();
                }
            },
            close: () => {
                dispatch({
                    type: Action.Close,
                });
            },
            getState: () => state,
        }));

        const hideKeyboardIfNeeded = () => {
            // TODO: Better condition handling, however, typo error appears in every combination
            if (
                searchInputRef &&
                (searchInputRef as RefObject<TextInput>).current
            ) {
                (searchInputRef as RefObject<TextInput>)?.current?.blur();
            }
        };

        const onPressOption: OnPressOptionType = (
            option: OptionType,
            optionIndex: number,
        ) => {
            if (closeDropdownOnSelect) {
                dispatch({ type: Action.Close });
            }

            const resolveOption = () => {
                if (!multiSelection) {
                    return {
                        selectedOption: option,
                        selectedOptionIndex: optionIndex,
                    };
                }

                const selectedOptionAsArray = selectedOption as
                    | OptionType[]
                    | null;
                const foundSelectedOption =
                    selectedOptionAsArray &&
                    selectedOptionAsArray.find(
                        (selectedOption: OptionType) =>
                            selectedOption.value === option.value,
                    );

                if (foundSelectedOption) {
                    return {
                        selectedOption: selectedOptionAsArray,
                        selectedOptionIndex:
                            typeof selectedOptionIndex === 'number'
                                ? selectedOptionIndex
                                : [...selectedOptionIndex],
                    };
                }

                const sOption = selectedOptionAsArray
                    ? selectedOptionAsArray.concat(option)
                    : [option];

                const sOptionIndex = optionsData
                    .map((item, index) => {
                        if (sOption.some(({ value }) => value === item.value)) {
                            return index;
                        }
                        return undefined;
                    })
                    .filter((item) => item !== undefined) as number[];

                return {
                    selectedOption: sOption,
                    selectedOptionIndex:
                        sOptionIndex.length > 0 ? [...sOptionIndex] : -1,
                };
            };

            dispatch({
                type: Action.SelectOption,
                payload: {
                    selectedOption: resolveOption().selectedOption,
                    selectedOptionIndex: resolveOption().selectedOptionIndex,
                },
            });

            if (searchable) {
                if (multiSelection) {
                    dispatch({ type: Action.SetSearchValue, payload: '' });
                } else {
                    dispatch({
                        type: Action.SetSearchValue,
                        payload: option.label,
                    });
                }
            }
            dispatch({ type: Action.SetOptionsData, payload: options });
            if (option) {
                hideKeyboardIfNeeded();
            }
        };

        const windowDimensions = useWindowDimensions();

        const setPosition = () => {
            if (containerRef.current) {
                containerRef.current.measure(
                    (_x, _y, width, height, pageX, pageY) => {
                        const listHeightFromProp =
                            StyleSheet.flatten(optionsListStyle)?.maxHeight;

                        const optionHeightFromProp =
                            StyleSheet.flatten(optionStyle)?.height;

                        const optionHeight = getSize({
                            size: optionHeightFromProp,
                            sizeType: 'height',
                            sizeFallback: ITEM_HEIGHT,
                            screenSize: windowDimensions.height,
                        });

                        const listHeight = getSize({
                            size: listHeightFromProp,
                            sizeType: 'height',
                            sizeFallback: MAX_HEIGHT_LIST,
                            screenSize: windowDimensions.height,
                        });

                        const finalHeight =
                            listHeight >= optionsData.length * optionHeight
                                ? optionsData.length * optionHeight
                                : listHeight;

                        const isOverflow =
                            pageY + height + finalHeight >
                            windowDimensions.height;

                        dispatch({
                            type: Action.SetPosition,
                            payload: {
                                width,
                                top: isOverflow
                                    ? pageY - finalHeight
                                    : pageY + height,
                                left: I18nManager.isRTL
                                    ? windowDimensions.width - width - pageX
                                    : pageX,
                                aboveSelectControl: isOverflow,
                            },
                        });
                    },
                );
            }
        };

        const onPressSelectControl: OnPressSelectControlType = () => {
            if (isOpened) {
                dispatch({
                    type: Action.Close,
                });
                return;
            }
            setPosition();
            if (containerRef.current) {
                dispatch({
                    type: Action.Open,
                });
            }
        };

        const onOutsidePress: OnOutsidePress = () => {
            dispatch({ type: Action.Close });
            dispatch({ type: Action.SetOptionsData, payload: options });
            if (searchable && selectedOptionTyped?.label) {
                dispatch({
                    type: Action.SetSearchValue,
                    payload: selectedOptionTyped.label,
                });
            }
            hideKeyboardIfNeeded();
        };

        useEffect(() => {
            if (isOpened) {
                onDropdownOpened?.();
            } else {
                onDropdownClosed?.();
            }
        }, [isOpened]);

        return (
            <View
                style={[styles.relative, selectContainerStyle]}
                onLayout={setPosition}
            >
                <SelectControl
                    ref={containerRef}
                    aboveSelectControl={aboveSelectControl}
                    animated={animated}
                    animationDuration={animationDuration}
                    clearable={clearable}
                    customLeftIconSource={customLeftIconSource}
                    customLeftIconStyles={customLeftIconStyles}
                    disabled={disabled}
                    dispatch={dispatch}
                    hideSelectControlArrow={hideSelectControlArrow}
                    isOpened={isOpened}
                    multiSelection={multiSelection}
                    multiSelectionOptionStyle={multiSelectionOptionStyle}
                    options={options}
                    placeholderText={placeholderText}
                    placeholderTextColor={placeholderTextColor}
                    searchPattern={searchPattern}
                    searchValue={searchValue}
                    searchable={searchable}
                    selectControlArrowImageStyle={selectControlArrowImageStyle}
                    selectControlButtonsContainerStyle={
                        selectControlButtonsContainerStyle
                    }
                    selectControlClearOptionA11yLabel={
                        selectControlClearOptionA11yLabel
                    }
                    selectControlClearOptionButtonHitSlop={
                        selectControlClearOptionButtonHitSlop
                    }
                    selectControlClearOptionButtonStyle={
                        selectControlClearOptionButtonStyle
                    }
                    selectControlClearOptionImageStyle={
                        selectControlClearOptionImageStyle
                    }
                    selectControlDisabledStyle={selectControlDisabledStyle}
                    selectControlOpenDropdownA11yLabel={
                        selectControlOpenDropdownA11yLabel
                    }
                    selectControlStyle={selectControlStyle}
                    selectControlTextStyle={selectControlTextStyle}
                    selectedOption={selectedOption}
                    selectedOptionIndex={selectedOptionIndex}
                    setPosition={setPosition}
                    onPressSelectControl={onPressSelectControl}
                    onSelect={onSelect}
                    inputProps={inputProps}
                />
                <OptionsList
                    NoOptionsComponent={NoOptionsComponent}
                    OptionComponent={OptionComponent}
                    aboveSelectControl={aboveSelectControl}
                    animated={animated}
                    animationDuration={animationDuration}
                    flatListProps={flatListProps}
                    isOpened={isOpened}
                    multiSelection={multiSelection}
                    noOptionsText={noOptionsText}
                    openedPosition={openedPosition}
                    optionSelectedStyle={optionSelectedStyle}
                    optionStyle={optionStyle}
                    optionTextStyle={optionTextStyle}
                    optionsData={optionsData}
                    optionsListStyle={optionsListStyle}
                    scrollToSelectedOption={scrollToSelectedOption}
                    searchValue={searchValue}
                    searchable={searchable}
                    searchedOptions={searchedOptions}
                    selectedOption={selectedOption}
                    selectedOptionIndex={selectedOptionIndex}
                    onOutsidePress={onOutsidePress}
                    onPressOption={onPressOption}
                    onSelect={onSelect}
                />
            </View>
        );
    },
);

type Styles = {
    relative: ViewStyle;
};

const styles = StyleSheet.create<Styles>({
    relative: {
        position: 'relative',
    },
});

Select.displayName = 'Select';
