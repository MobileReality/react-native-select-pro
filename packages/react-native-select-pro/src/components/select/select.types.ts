import type { Dispatch, ForwardedRef, RefObject } from 'react';
import type { View } from 'react-native';

import type { ActionType, State } from '../../state';
import type { OptionalToRequired, SelectProps, SelectRef } from '../../types';

type FromSelectComponent<T> = Pick<
    SelectProps<T>,
    | 'defaultOption'
    | 'disabled'
    | 'searchable'
    | 'multiple'
    | 'closeOptionsListOnSelect'
    | 'onSelectOpened'
    | 'onSelectClosed'
    | 'onRemove'
    | 'onSectionSelect'
    | 'onSectionRemove'
    | 'onSelect'
>;

export type UseSelect<T> = OptionalToRequired<
    FromSelectComponent<T> & {
        dispatch: Dispatch<ActionType<T>>;
        selectControlRef: RefObject<View>;
        optionsListRef: RefObject<View>;
        ref: ForwardedRef<SelectRef<T>>;
        state: State<T>;
    }
>;
