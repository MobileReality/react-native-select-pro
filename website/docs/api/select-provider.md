---
id: select-provider
title: SelectProvider
sidebar_label: SelectProvider
---

`<SelectProvider>` must wrap a whole app, so should be placed in the root app component file (index.js, App.js or sth like that). So children of this component is one required prop.

### children
```typescript jsx
children: React.ReactNode;
```

#### Example

```jsx
import React from 'react';
import { SelectProvider } from '@mobile-reality/react-native-select-pro';

const RootComponent = () => {
  return (
    <SelectProvider>
        {/* rest of your app code */}
    </SelectProvider>
  )
};
```
