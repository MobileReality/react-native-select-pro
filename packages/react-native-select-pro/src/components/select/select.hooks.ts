import { useCallback, useContext, useEffect, useImperativeHandle } from 'react';
import type { LayoutRectangle } from 'react-native';
import { I18nManager, useWindowDimensions } from 'react-native';

import { APPROX_STATUSBAR_HEIGHT } from '../../constants';
import { isValidDefaultOption } from '../../helpers';
import {
    getReducedSectionData,
    getSectionOptionsIndexes,
    selectedOptionResolver,
} from '../../helpers';
import { Action } from '../../state';
import type {
    OnOutsidePress,
    OnPressOptionType,
    OnPressSelectControlType,
    OptionType,
    SelectRef,
} from '../../types';
import { isSectionOptionsType } from '../../types';
import { SelectModalContext } from '../select-provider';

import type { UseSelect } from './select.types';

export const useSelect = <T>({
    ref,
    selectControlRef,
    state,
    defaultOption,
    disabled,
    closeOptionsListOnSelect,
    searchable,
    multiple,
    dispatch,
    onRemove,
    onSelectOpened,
    onSelectClosed,
    onSectionSelect,
    onSectionRemove,
    optionsListRef,
    onSelect,
}: UseSelect<T>) => {
    const { height: screenHeight, width: screenWidth } = useWindowDimensions();
    const valueY = useContext(SelectModalContext);
    const { isOpened, selectedOption, optionsData, searchInputRef, selectedOptionIndex } = state;
    const { selectedOptionLabel, selectedOptions } = selectedOptionResolver(selectedOption);
    const isSectionedOptions = isSectionOptionsType(optionsData);

    const open = useCallback(() => {
        dispatch({ type: Action.Open });
    }, [dispatch]);

    const close = useCallback(() => {
        dispatch({ type: Action.Close });
    }, [dispatch]);

    useEffect(() => {
        const setDefaultOption = () => {
            const isValidPassDefaultOption = isValidDefaultOption(defaultOption);

            if (optionsData.length === 0 || !isValidPassDefaultOption) {
                return;
            }

            const foundIndex = isSectionedOptions
                ? getReducedSectionData(optionsData).indexOf(defaultOption)
                : optionsData.indexOf(defaultOption);

            dispatch({
                type: Action.SelectOption,
                payload: {
                    selectedOption: defaultOption,
                    selectedOptionIndex: foundIndex,
                },
            });
        };

        setDefaultOption();
    }, [optionsData, defaultOption, isSectionedOptions, dispatch]);

    const measureSelectInWindow = () => {
        return new Promise<LayoutRectangle>((resolve) => {
            if (selectControlRef.current) {
                selectControlRef.current.measureInWindow((x, y, width, height) => {
                    resolve({ x, y, width, height });
                });
            }
        });
    };

    const measureOptionsListInWindow = () => {
        return new Promise<Pick<LayoutRectangle, 'height'>>((resolve) => {
            if (optionsListRef.current) {
                optionsListRef.current.measureInWindow((_x, _y, _width, height) => {
                    resolve({ height });
                });
            }
        });
    };

    const setOptionsListPosition = async () => {
        const { x, y, width, height } = await measureSelectInWindow();
        const { height: optionsListHeight } = await measureOptionsListInWindow();

        const isOverflow = y + height + optionsListHeight > screenHeight;
        const top = isOverflow
            ? y - optionsListHeight
            : y + height - valueY + APPROX_STATUSBAR_HEIGHT;
        const left = I18nManager.getConstants().isRTL ? screenWidth - width - x : x;

        dispatch({
            type: Action.SetOptionsListPosition,
            payload: {
                width,
                top,
                left,
                aboveSelectControl: isOverflow,
            },
        });
    };

    const onPressSelectControl: OnPressSelectControlType = async () => {
        if (disabled) {
            return;
        }

        if (isOpened) {
            close();
            return;
        }

        if (selectControlRef.current) {
            open();
            await setOptionsListPosition();
        }
    };

    useImperativeHandle(
        ref,
        (): SelectRef<T> => ({
            clear: () => {
                dispatch({
                    type: Action.SelectOption,
                    payload: { selectedOption: null, selectedOptionIndex: -1 },
                });
                if (onRemove && selectedOption) {
                    onRemove(selectedOption, selectedOptionIndex);
                }
            },
            open: async () => {
                if (selectControlRef.current) {
                    open();
                    await setOptionsListPosition();
                }
            },
            close,
            getState: () => state,
        }),
    );

    const hideKeyboardIfNeeded = useCallback(
        () => searchInputRef?.current?.blur(),
        [searchInputRef],
    );

    const resolveOption = useCallback(
        (option: OptionType<T>, optionIndex: number) => {
            if (!multiple) {
                return {
                    selectedOption: option,
                    selectedOptionIndex: optionIndex,
                };
            }

            const foundSelectedOption = selectedOptions?.find(
                (selectedOption: OptionType<T>) => selectedOption.value === option.value,
            );
            if (foundSelectedOption) {
                return {
                    selectedOption: selectedOptions,
                    selectedOptionIndex:
                        typeof selectedOptionIndex === 'number'
                            ? selectedOptionIndex
                            : [...selectedOptionIndex],
                };
            }

            const mergedOptions = selectedOptions ? selectedOptions.concat(option) : [option];
            const resolvedOptionsData = isSectionedOptions
                ? getReducedSectionData(optionsData)
                : optionsData;
            const optionsIndexes = mergedOptions
                .map((item) => resolvedOptionsData.findIndex(({ value }) => value === item.value))
                .filter((item): item is number => item !== null);

            return {
                selectedOption: mergedOptions,
                selectedOptionIndex: optionsIndexes.length > 0 ? [...optionsIndexes] : -1,
            };
        },
        [isSectionedOptions, multiple, optionsData, selectedOptionIndex, selectedOptions],
    );

    const onPressOption: OnPressOptionType<T> = useCallback(
        (option: OptionType<T>, optionIndex: number) => {
            if (closeOptionsListOnSelect) {
                close();
            }

            dispatch({
                type: Action.SelectOption,
                payload: resolveOption(option, optionIndex),
            });

            if (searchable) {
                if (multiple) {
                    dispatch({ type: Action.SetSearchValue, payload: '' });
                } else {
                    dispatch({
                        type: Action.SetSearchValue,
                        payload: option.label,
                    });
                }
            }

            if (option) {
                hideKeyboardIfNeeded();
            }

            // callback
            if (onSelect) {
                onSelect(option, optionIndex);
            }
        },
        [
            close,
            closeOptionsListOnSelect,
            dispatch,
            hideKeyboardIfNeeded,
            multiple,
            onSelect,
            resolveOption,
            searchable,
        ],
    );

    const onPressSection = (title: string) => {
        if (closeOptionsListOnSelect && multiple) {
            close();
        }

        if (!multiple || !isSectionedOptions) {
            return;
        }

        const resolveOption = () => {
            const sectionOptions = optionsData.find((item) => item.title === title)?.data ?? [];
            const formattedSectionOptions = sectionOptions.map((item) => ({
                ...item,
                section: {
                    title,
                    index: optionsData.findIndex((el) => el.title === title),
                },
            }));
            const newSectionOptions = formattedSectionOptions.filter(
                (item) => !selectedOptions?.some((selected) => selected.value === item.value),
            );

            if (newSectionOptions.length === 0 && selectedOptions) {
                onSectionRemove?.(
                    formattedSectionOptions,
                    getSectionOptionsIndexes(optionsData, formattedSectionOptions),
                );
                const restOptions = selectedOptions.filter(
                    (item) =>
                        !formattedSectionOptions.some((selected) => selected.value === item.value),
                );

                return {
                    selectedOption: restOptions.length > 0 ? restOptions : null,
                    selectedOptionIndex: restOptions
                        ? getSectionOptionsIndexes(optionsData, restOptions)
                        : -1,
                };
            }

            const mergedOptions = selectedOptions
                ? selectedOptions.concat(newSectionOptions)
                : newSectionOptions;

            const optionsIndexes = getSectionOptionsIndexes(optionsData, mergedOptions);

            onSectionSelect?.(
                newSectionOptions,
                getSectionOptionsIndexes(optionsData, newSectionOptions),
            );

            return {
                selectedOption: mergedOptions,
                selectedOptionIndex: optionsIndexes.length > 0 ? optionsIndexes : -1,
            };
        };

        dispatch({
            type: Action.SelectOption,
            payload: resolveOption(),
        });

        if (searchable) {
            dispatch({ type: Action.SetSearchValue, payload: '' });
        }
    };

    const onOutsidePress: OnOutsidePress = () => {
        if (selectedOptionLabel && searchable) {
            dispatch({
                type: Action.SetSearchValue,
                payload: selectedOptionLabel,
            });
        }

        close();
        hideKeyboardIfNeeded();
    };

    useEffect(() => {
        if (isOpened) {
            onSelectOpened?.();
        } else {
            onSelectClosed?.();
        }
    }, [isOpened, onSelectOpened, onSelectClosed]);

    return {
        setOptionsListPosition,
        onPressOption,
        onPressSection,
        onOutsidePress,
        onPressSelectControl,
    };
};
