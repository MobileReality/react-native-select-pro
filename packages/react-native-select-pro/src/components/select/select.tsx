import type { Dispatch, ForwardedRef, Reducer } from 'react';
import React, { forwardRef, useReducer, useRef } from 'react';
import type { SectionListData, ViewStyle } from 'react-native';
import { StyleSheet, UIManager, View } from 'react-native';
import { Portal } from '@gorhom/portal';

import { COLORS, Portals } from '../../constants';
import { OptionsListContextProvider, SelectContextProvider } from '../../context';
import { isAndroid, mergeObjects } from '../../helpers';
import type { ActionType, CreateInitialStateType, State } from '../../state';
import { createInitialState, reducer } from '../../state';
import { themes } from '../../themes';
import type {
    OnPressOptionType,
    OptionsType,
    OptionType,
    SelectProps,
    SelectRef,
    SelectStyles,
} from '../../types';
import { Backdrop } from '../backdrop';
import { OptionsList } from '../options-list';
import { SelectControl } from '../select-control';

import { useSelect } from './select.hooks';

if (isAndroid && UIManager.setLayoutAnimationEnabledExperimental) {
    UIManager.setLayoutAnimationEnabledExperimental(true);
}

const SelectComponent = <T,>(props: SelectProps<T>, ref: ForwardedRef<SelectRef<T>>) => {
    const {
        // Required
        options,
        // Basic
        animation = true,
        clearable = true,
        closeOptionsListOnSelect = true,
        defaultOption,
        disabled = false,
        hasBackdrop = true,
        hideArrow = false,
        multiple = false,
        noOptionsText = 'No options',
        placeholderText = 'Select...',
        placeholderTextColor = COLORS.GRAY,
        pressableSelectedOption = true,
        scrollToSelectedOption = true,
        searchable = false,
        searchPattern = (payload: string) => `(${payload})`,
        styles: customStyles,
        theme = 'none',
        // Callbacks
        onSelectChangeText,
        onSectionSelect,
        onSectionRemove,
        onSelect,
        onSelectOpened,
        onSelectClosed,
        onRemove,
        // Customized
        arrowContainerProps,
        arrowImageProps,
        backdropChildProps,
        backdropProps,
        clearOptionButtonProps,
        clearOptionImageProps,
        flatListProps,
        noOptionsProps,
        noOptionsTextProps,
        optionButtonProps,
        optionTextProps,
        selectContainerProps,
        selectInputProps,
        selectLeftIconImageProps,
        selectLeftIconsProps,
        selectRightIconsProps,
        selectTextProps,
        sectionHeaderButtonProps,
        sectionHeaderImageProps,
        sectionHeaderTextProps,
        sectionListProps,
    } = props;

    const [state, dispatch] = useReducer<
        Reducer<State<T>, ActionType<T>>,
        CreateInitialStateType<T>
    >(reducer, { options, searchable, animation }, createInitialState);

    const {
        isOpened,
        selectedOption,
        optionsData,
        openedPosition,
        searchValue,
        searchedOptions,
        selectedOptionIndex,
    } = state;

    const { aboveSelectControl } = openedPosition;
    const mainStyles: SelectStyles = mergeObjects(themes[theme], customStyles);

    const selectControlRef = useRef<View>(null);
    const optionsListRef = useRef<View>(null);

    const {
        setOptionsListPosition,
        onPressOption,
        onOutsidePress,
        onPressSelectControl,
        onPressSection,
    } = useSelect<T>({
        selectControlRef,
        optionsListRef,
        dispatch,
        defaultOption,
        onRemove,
        disabled,
        closeOptionsListOnSelect,
        searchable,
        multiple,
        onSelectOpened,
        onSelectClosed,
        ref,
        state,
        onSectionSelect,
        onSectionRemove,
        onSelect,
    });

    return (
        <View style={styles.relative}>
            <SelectContextProvider
                value={{
                    isOpened,
                    animation,
                    aboveSelectControl,
                    clearable,
                    disabled,
                    hideArrow,
                    multiple,
                    optionsData: optionsData as OptionsType<unknown>,
                    placeholderText,
                    placeholderTextColor,
                    searchPattern,
                    searchValue,
                    onPressSelectControl,
                    selectInputProps,
                    onRemove,
                    dispatch: dispatch as Dispatch<ActionType<unknown>>,
                    setOptionsListPosition,
                    selectedOption,
                    selectedOptionIndex,
                    styles: mainStyles,
                    clearOptionButtonProps,
                    clearOptionImageProps,
                    arrowContainerProps,
                    arrowImageProps,
                    selectRightIconsProps,
                    selectLeftIconsProps,
                    selectLeftIconImageProps,
                    selectTextProps,
                    selectContainerProps,
                    onSelectChangeText,
                }}
            >
                <SelectControl ref={selectControlRef} />
            </SelectContextProvider>
            {isOpened && (
                <>
                    {hasBackdrop && (
                        <Portal hostName={Portals.Backdrop}>
                            <Backdrop
                                backdropCustomStyles={mainStyles?.backdrop}
                                backdropProps={backdropProps}
                                backdropChildProps={backdropChildProps}
                                onOutsidePress={onOutsidePress}
                            />
                        </Portal>
                    )}
                    <Portal hostName={Portals.OptionsList}>
                        <OptionsListContextProvider
                            value={{
                                animation,
                                aboveSelectControl,
                                flatListProps,
                                isOpened,
                                noOptionsText,
                                openedPosition,
                                optionsData: optionsData as OptionsType<unknown>,
                                scrollToSelectedOption,
                                searchValue,
                                onPressOption: onPressOption as OnPressOptionType<unknown>,
                                onPressSection,
                                selectedOption,
                                searchedOptions: searchedOptions as OptionsType<unknown>,
                                selectedOptionIndex,
                                sectionListProps: sectionListProps as SectionListData<
                                    OptionsType<OptionType>
                                >,
                                styles: mainStyles,
                                optionButtonProps,
                                optionTextProps,
                                noOptionsProps,
                                noOptionsTextProps,
                                sectionHeaderButtonProps,
                                sectionHeaderImageProps,
                                sectionHeaderTextProps,
                                pressableSelectedOption,
                                multiple,
                                disabled,
                            }}
                        >
                            <OptionsList ref={optionsListRef} />
                        </OptionsListContextProvider>
                    </Portal>
                </>
            )}
        </View>
    );
};

type Styles = {
    relative: ViewStyle;
};

const styles = StyleSheet.create<Styles>({
    relative: {
        position: 'relative',
    },
});

export const Select = forwardRef(SelectComponent) as <T>(
    props: SelectProps<T> & { ref?: ForwardedRef<SelectRef<T>> },
) => ReturnType<typeof SelectComponent>;
