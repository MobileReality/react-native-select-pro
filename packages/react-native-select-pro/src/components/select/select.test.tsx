import React from 'react';
import { render } from '@testing-library/react-native';

import { SelectProvider } from '../select-provider';

import { Select } from './select';

describe('Select', () => {
    it('should generate Select snapshot', () => {
        const DATA = [{ label: 'test', value: 'test' }];
        const select = render(
            <SelectProvider>
                <Select options={DATA} />
            </SelectProvider>,
        );
        expect(select).toMatchSnapshot();
    });
});
