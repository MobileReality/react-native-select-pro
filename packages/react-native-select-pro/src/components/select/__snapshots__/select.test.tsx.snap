// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select should generate Select snapshot 1`] = `
[
  <View
    onLayout={[Function]}
    style={
      [
        {
          "position": "relative",
        },
        undefined,
      ]
    }
  >
    <View
      style={
        {
          "position": "relative",
        }
      }
    >
      <View
        accessibilityLabel="Open a dropdown"
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "backgroundColor": "#ffffff",
              "borderRadius": 4,
              "borderWidth": 1,
              "flexDirection": "row",
              "height": 40,
            },
            false,
            undefined,
            {},
          ]
        }
      >
        <View
          style={
            [
              {
                "flex": 1,
                "height": "100%",
                "justifyContent": "center",
                "paddingHorizontal": 12,
              },
              {
                "paddingRight": 55,
              },
            ]
          }
        >
          <Text
            numberOfLines={1}
            style={
              [
                {
                  "fontSize": 12,
                  "textAlign": "left",
                },
                undefined,
                {
                  "color": "#808080",
                },
              ]
            }
          >
            Select...
          </Text>
        </View>
        <View
          style={
            [
              {
                "alignItems": "center",
                "flexDirection": "row",
                "height": "100%",
                "justifyContent": "center",
                "position": "absolute",
                "right": 8,
                "top": 0,
              },
              undefined,
            ]
          }
        >
          <Image
            collapsable={false}
            source={
              {
                "testUri": "packages/react-native-select-pro/src/assets/icons/chevron-down.png",
              }
            }
            style={
              {
                "height": 25,
                "transform": [
                  {
                    "rotate": "0deg",
                  },
                ],
                "width": 25,
                "zIndex": -1,
              }
            }
          />
        </View>
      </View>
    </View>
  </View>,
  <View
    collapsable={false}
    pointerEvents="none"
    style={
      {
        "backgroundColor": "#ffffff",
        "borderBottomLeftRadius": 4,
        "borderBottomRightRadius": 4,
        "borderTopWidth": 0,
        "borderWidth": 1,
        "elevation": 5,
        "flex": 1,
        "left": 0,
        "maxHeight": 120,
        "opacity": 0,
        "position": "absolute",
        "top": 0,
        "width": 0,
        "zIndex": 1,
      }
    }
  >
    <RCTScrollView
      ListEmptyComponent={
        <NoOptions
          noOptionsText="No options"
        />
      }
      accessibilityLabel="Options list"
      accessibilityState={
        {
          "expanded": false,
        }
      }
      bounces={false}
      data={
        [
          {
            "label": "test",
            "value": "test",
          },
        ]
      }
      getItem={[Function]}
      getItemCount={[Function]}
      getItemLayout={[Function]}
      keyExtractor={[Function]}
      keyboardShouldPersistTaps="handled"
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      persistentScrollbar={true}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={[]}
      testID="Options list"
      viewabilityConfigCallbackPairs={[]}
    >
      <View>
        <View
          style={null}
        >
          <View
            accessibilityLabel="Choose test option"
            accessibilityRole="button"
            accessibilityState={
              {
                "disabled": false,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            forwardedRef={[Function]}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "height": 40,
                "justifyContent": "center",
                "opacity": 1,
                "paddingHorizontal": 12,
              }
            }
          >
            <Text
              numberOfLines={1}
              style={
                [
                  {
                    "color": "#000000",
                    "fontSize": 12,
                    "textAlign": "left",
                  },
                  undefined,
                ]
              }
            >
              test
            </Text>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>,
]
`;
